### PR TITLE
[BUG] ClusterLoadFallbackPolicy is not strictness when a shuffle with big partitions to register

### PR DESCRIPTION
**Pull request by TonyDoen**
# [BUG] ClusterLoadFallbackPolicy is not strictness when a shuffle with big partitions to register

### What changes were proposed in this pull request?
If a shuffle with 5w partitions to register, and cluster has 3w but the slots usage is not in high load status, 5w allocation slots will failed.

### Why are the changes needed?
This change will avoid this circumstance that a shuffle that it will cost too much slots over clusters' continues process of RSS handling

### Related issues.
https://github.com/alibaba/RemoteShuffleService/issues/17

### How was this patch tested?

/cc @FMX

/assign @wangshengjie

### CODE REVIEW

File: client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleFallbackPolicyRunner.scala [hunk 1]
```diff
@@ -31,7 +31,8 @@ class RssShuffleFallbackPolicyRunner(sparkConf: SparkConf) extends Logging {
   def applyAllFallbackPolicy(dependency: ShuffleDependency[_, _, _],
```
Comment (wangshengjie123) [inline client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleFallbackPolicyRunner.scala:31]: `dependency.partitioner.numPartitions` seems enough, could you please help to optimize this？Thanks

File: server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala [hunk 2]
```diff
@@ -421,12 +421,12 @@ private[deploy] class Master(
     context.reply(OneWayMessageResponse)
   }
 
-  private def handleGetClusterLoadStatus(context: RpcCallContext): Unit = {
-    val (_, _, _, result) = getClusterLoad
+  private def handleGetClusterLoadStatus(context: RpcCallContext, numPartitions: Int): Unit = {
+    val (_, _, _, result) = getClusterLoad(numPartitions)
```
Comment (wangshengjie123) [inline server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala:425]: Why we need so much return values, please remove useless return

Comment (TonyDoen) [reply server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala:425]: the [getClusterLoad] is common function, we may need the top three return-values in the other place. I can separate the value through creating new function.

File: client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleFallbackPolicyRunner.scala (hunk 1 above)
Comment (TonyDoen) [reply client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleFallbackPolicyRunner.scala]: Yes, I have retrieved the param [dependency] from this function

**Comment by FMX** [comment]
LGTM

File: server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala [hunk 3]
```diff
@@ -421,32 +421,34 @@ private[deploy] class Master(
...
+    val totalUsedRatio: Double = (usedSlots + numPartitions) / totalSlots.toDouble
+    val result = totalUsedRatio >= clusterSlotsUsageLimit
+    logInfo(s"Current cluster slots usage:$totalUsedRatio, conf:$clusterSlotsUsageLimit, " +
+        s"overload:$result")
     context.reply(GetClusterLoadStatusResponse(result))
   }
 
-  private def getClusterLoad: (Int, Int, Int, Boolean) = {
+  private def getClusterLoad(numPartitions: Int = 0): (Int, Int, Int) = {
     if (workersSnapShot.isEmpty) {
```
Comment (waitinfuture) [inline server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala:436]: workersSnapShot is a method and will return a new object each invocation. I suggest to assign to a variable here.

File: client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleFallbackPolicyRunner.scala [hunk 4]
```diff
@@ -75,8 +74,9 @@ class RssShuffleFallbackPolicyRunner(sparkConf: SparkConf) extends Logging {
    * if rss cluster is under high load, fallback to external shuffle
    * @return if rss cluster's slots used percent is overhead the limit
    */
-  def applyClusterLoadFallbackPolicy(lifeCycleManager: LifecycleManager): Boolean = {
-    val needFallback = lifeCycleManager.isClusterOverload()
+  def applyClusterLoadFallbackPolicy(lifeCycleManager: LifecycleManager, numPartitions: Int):
```
Comment (waitinfuture) [inline client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleFallbackPolicyRunner.scala:77]: We'd better add a switch config to indicate whether to enable load check, like
```
rss.clusterload.fallback.enabled
```

File: server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala (hunk 3 above)
Comment (TonyDoen) [reply server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala]: Thanks, I have retrieved a variable [workers] keeping [workersSnapShot] result

**Pull request merged**